diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
Fix for a regression in the management of the FFTW library
https://i10git.cs.fau.de/walberla/walberla/-/issues/191
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -29,7 +29,7 @@ if ( CMAKE_CUDA_COMPILER )
 endif()
 add_subdirectory( domain_decomposition )
 add_subdirectory( executiontree )
-if ( FFTW3_FOUND )
+if ( WALBERLA_BUILD_WITH_FFTW AND FFTW3_FOUND )
    add_subdirectory( fft )
 endif()
 add_subdirectory( field )
