https://i10git.cs.fau.de/walberla/walberla/-/merge_requests/709
--- src/core/mpi/RecvBuffer.h
+++ src/core/mpi/RecvBuffer.h
@@ -230 +230 @@
-   sb.begin_ = new T[0];
+   sb.begin_ = nullptr;
--- src/core/mpi/SendBuffer.h
+++ src/core/mpi/SendBuffer.h
@@ -700 +700 @@
-   begin_ = new T[0];
+   begin_ = nullptr;
support for CMake 4.0
--- cmake/FindOpenMesh.cmake
+++ cmake/FindOpenMesh.cmake
@@ -56 +56 @@
-cmake_minimum_required(VERSION 3.5.0)
+cmake_minimum_required(VERSION 3.10.0)
